rfc/rfc-0013-invoker-tck.md

```diff
@@ -0,0 +1,37 @@
+# RFC-0013: Invoker TCK
+
+**Authors:** Eric Bottard
+
+**Status:** Accepted
+
+**Pull Request URL:** https://github.com/projectriff/riff/pull/1378
+
+**Superseded by:** N/A
+
+**Supersedes:** N/A
+
+**Related:** 
+
+
+## Problem
+A specification of how function invokers should produce a runnable function exists (available at https://github.com/projectriff/invoker-specification). While it does its best to be prescriptive about what an invoker should and should not do, there are subtle caveats about the expected behavior, in particular in the _streaming_ case. Having a compatibility kit that implementors can run to ensure compliance would help detecting problems upstream.
+
+## Solution
+This RFC proposes to introduce a TCK that would take the form of at least a runnable tool (maybe also available as a go library) that focuses on testing the conformance of packaged functions with the _interaction (networking) protocols_ prescribed by the invoker specification.
+As such, it would require that users of the TCK produce a set of Docker images implementing a handful of trivial functions (such as a "square" function, see below). 
+How these images got constructed is deemed out of scope of the TCK, and while riff and the specification propose using buildpacks to construct such images, the TCK focuses on the resulting behavior of the container, which is assumed to result from the usage of the _invoker layer_.
+The use of containers as the unit of testing has several advantages:
+- the overall environment needed for the invoker to function properly does not have to be set up as part of the templated run of a test. Invokers implementors are in full control to provide a runnable container, and there is no software conflict with the environment used to drive the TCK run.
+- using a container guarantees isolation of test runs, in particular with regard to data left behind.
+
+The TCK would roughly take the following form:
+A series of test runs that cover the actual requirements of the specification. In particular, the words "MUST" throughout the spec should correspond to actual testcases. To run such tests, the invoker implementor is asked to provide a series of containers that implement simple functions. For example, to test the http request/reply interaction and correct support for marshalling/unmarshalling of `application/json`, the TCK may leverage a container that assumes input as a number and returns its square. The TCK runner would run the container image configured for "square", POST a payload on `http://<host>:8080/` and verify synchronous return of a squared number, with `Content-Type` compatible with the requested `Accept` header.
```

Comment:
How will the TCK test streaming functions which consume input but which produce no output? I would have thought it would be necessary to inject a remote endpoint into the user-provided function so that the function can say what it has received. Without this, it will be hard to know that an invoker isn't somehow buffering input inappropriately.

Reply:
Consumers are not really supported at the moment (may work with Node, does not work with Java).
I would expect consumers and suppliers to be out of scope of this RFC.
There should be a subsequent RFC to evolve both the programmatic model of invokers and the TCK, in my opinion.

Reply:
> How will the TCK test streaming functions which consume input but which produce no output?

StdOut/Err are always available

> Consumers are not really supported at the moment

That sounds like a bug rather than an intentional design. Do we have issues?

A Processor currently requires at least one input stream and no output streams.

Reply:
> A Processor currently requires at least one input stream and no output streams.

Reframing: using a `java.util.function.Consumer` is not supported at the moment, which would be the idiomatic way to express the kind of functions you describe. I personally have not tested this kind of functions, FATS does not either it seems.